pkg/lexer/lexer.go

```diff
@@ -0,0 +1,84 @@
+package lexer
+
+type TokenType int
+
+const (
+	// Special tokens
+	ILLEGAL TokenType = iota
+	EOF
+	SEMICOLON // ;
+	NEWLINE   // \n
+
+	// Identifiers and literals
+	IDENT
+
+	// Logical operators
+	AND
+	OR
+
+	// Quote operators
+	SINGLE_QUOTE
+	DOUBLE_QUOTE
+
+	// Redirection operators
+	TO_RIGHT
+	TO_LEFT
+
+	// Pipe operators
+	PIPE
+)
+
+func (t TokenType) String() string {
+	var tokenMap = map[TokenType]string{
+		EOF:          "EOF",
+		SEMICOLON:    ";",
+		NEWLINE:      "\n",
+		IDENT:        "IDENT",
+		AND:          "&&",
+		OR:           "||",
+		SINGLE_QUOTE: "'",
+		DOUBLE_QUOTE: "\"",
+		TO_RIGHT:     ">",
+		TO_LEFT:      "<",
+		PIPE:         "|",
+	}
+
+	return tokenMap[t]
+}
+
+type Token struct {
+	Type  TokenType
+	Value string
+}
+
+// StateFn this is a recursive definition of the State - makes it easy to navigate ast type structures.
+type StateFn func(*Lexer) StateFn
+
+type Lexer struct {
+	Name   string
+	Input  string
+	State  StateFn
+	Tokens chan Token
+	Start  int
+	Pos    int
+	End    int
```

Comment:
And a line and column field here, for reporting the position.

Reply:
I think line field makes sense here but not sure about the column field

Reply:
How do you intend to find the current column?

Reply:
We can't but thats exactly my question. Why do we need the column?

Reply:
Normally, good error reports contain both the line and column. We can also use it later to do things like:
```
15 | console.log(obj,)
                    ^
```
Of course, we can ignore the column, but sometimes, you have lines like:
```
15 | /(https?:\/\/(?:www\.|(?!www))[^\s.]+\.[^\s]{2,}|www\.[^\s]+\.[^\s]{2,})/gi
```
Which are hard to read, and mentioning the column helps a lot here.

The above example is from my own experience. there was a problem with a `\s`. There were many `\s`s and mentioning the column in the error saved a lot of trial and error.

Reply:
Are we creating a new language or a variant of shell? Shell is able to run binaries and redirect and pipe right and that's it right? If a command fails inside we can't really do anything because the error reporting is done by the binary itself. If a pipe or redirect fails we would have an error reporting but I have no idea how useful that would be. 

```sh
15 | /(https?:\/\/(?:www\.|(?!www))[^\s.]+\.[^\s]{2,}|www\.[^\s]+\.[^\s]{2,})/gi
```
Is the above example a use case in shell? That is just a reg exp and mostly used by grep and other binaries not by shell itself (at least I haven't come across shells utility to handle regex) so ... 

In any case, I can add it so you can just make a decision on this and I'll do it.

Reply:
Lets not implement in right now then. Later, if we fell it is needed, we can implement it.

Reply:
Cool, I'll make my changes in the next two, three days - I will reduce the exposed API footprint quite a bit and write the parser in the same package (later will change the package name to parser)

Reply:
I have added the line field but the column field can be inferred to at runtime from the pos. But I've also made it quite easy to add that information (similar to how I just added the line and pos information) so can be added later.